runtime/compiler/env/PersistentAllocator.cpp

```diff
@@ -228,58 +324,256 @@ void
 PersistentAllocator::freeFixedSizeBlock(Block * block)
    {
    // Appropriate lock should have been obtained
-   TR_ASSERT(block->_size > 0, "Block size is non-positive");
-   size_t const index = freeBlocksIndex(block->_size);
+   TR_ASSERT(block->size() > 0, "Block size is non-positive");
+   size_t const index = freeBlocksIndex(block->size());
    TR_ASSERT(index != LARGE_BLOCK_LIST_INDEX, "freeFixedSizeBlock should be used for small blocks, so index cannot be LARGE_BLOCK_LIST_INDEX");
-   block->_next = _freeBlocks[index];
+   block->setNext(_freeBlocks[index]);
    _freeBlocks[index] = block;
    }
 
 void
 PersistentAllocator::freeVariableSizeBlock(Block * block)
    {
    // Appropriate lock should have been obtained
-   TR_ASSERT(block->_size > 0, "Block size is non-positive");
-   block->_next = NULL;
+   TR_ASSERT(block->size() > 0, "Block size is non-positive");
+   block->setNext(NULL);
    // Add block to the variable-size-block chain which is in ascending size order.
    //
-   TR_ASSERT(freeBlocksIndex(block->_size) == LARGE_BLOCK_LIST_INDEX, "freeVariableSizeBlock should be used for large blocks, so index should be LARGE_BLOCK_LIST_INDEX");
+   TR_ASSERT(freeBlocksIndex(block->size()) == LARGE_BLOCK_LIST_INDEX, "freeVariableSizeBlock should be used for large blocks, so index should be LARGE_BLOCK_LIST_INDEX");
    Block * blockIterator = _freeBlocks[LARGE_BLOCK_LIST_INDEX];
-   if (!blockIterator || !(blockIterator->_size < block->_size) )
+   if (!blockIterator || !(blockIterator->size() < block->size()) )
       {
-      block->_next = _freeBlocks[LARGE_BLOCK_LIST_INDEX];
+      block->setNext(_freeBlocks[LARGE_BLOCK_LIST_INDEX]);
       _freeBlocks[LARGE_BLOCK_LIST_INDEX] = block;
       }
    else
       {
-      while (blockIterator->next() && blockIterator->next()->_size < block->_size)
+      while (blockIterator->next() && blockIterator->next()->size() < block->size())
          {
          blockIterator = blockIterator->next();
          }
-      block->_next = blockIterator->next();
-      blockIterator->_next = block;
+      block->setNext(blockIterator->next());
+      blockIterator->setNext(block);
+      }
+   }
+
+#if defined(J9VM_OPT_JITSERVER)
+size_t 
+PersistentAllocator::getInterval(size_t blockSize)
+   {  
+   // Find the power-of-two interval that this block size belongs to
+   TR_ASSERT(blockSize >= PERSISTANT_BLOCK_SIZE_BUCKETS * sizeof(void *), "getInterval should be used only on big blocks. blockSize=%zu", blockSize);
+   // If very large block
+   if (blockSize >= (1 << (BITS_TO_SHIFT_FIRST_INTERVAL + NUM_INTERVALS - 1)))
+      return NUM_INTERVALS - 1; // last one
+   blockSize >>= (BITS_TO_SHIFT_FIRST_INTERVAL + 1);
```

Comment:
You are right, `blockSize` needs to be shifted only by `BITS_TO_SHIFT_FIRST_INTERVAL`.
The original code for finding the interval is commented out at the end of this function. With that code I needed to shift by `BITS_TO_SHIFT_FIRST_INTERVAL+1`, but with my new (and supposedly faster) implementation using the lookup table I need to delete that 1.

Reply:
Regarding `if (blockSize >= (1 << (BITS_TO_SHIFT_FIRST_INTERVAL + NUM_INTERVALS - 1)))`, did you still want to max it 2^14? I believe 2^15 - 1 would yield a shift of 7, so you can make the maximum 2^16 (ie `BITS_TO_SHIFT_FIRST_INTERVAL + NUM_INTERVALS + 1`)

Reply:
I could increase the number of intervals from 8 to 9, though, from experiments, the number of blocks with very large sizes is low. Maybe just to be future proof.

Reply:
Sorry, I didn't mean to increase the number of intervals; I meant that with the current 8 intervals, `BITS_TO_SHIFT_FIRST_INTERVAL + NUM_INTERVALS - 1 == 14` whereas you can have it be `BITS_TO_SHIFT_FIRST_INTERVAL + NUM_INTERVALS + 1` with the 8 intervals.

Reply:
since 2^15 - 1 will yield a `0xFFFFFFFF` which, when shifted by 8 will give you `0xFFFF`; the upper 4 bits will then be 15, so it can be indexed into the `_logTable` array.

Reply:
I don't understand what this will achieve. Currently, if the block size is greater than 16K, I directly return interval=7 (last one) without any log2 function (thus saving unnecessary overhead)

Reply:
I guess right now, 2^14 - 1 would get you 0x3FFF; after rsh8, you get 0x3F; so the interval you'd get out of this method is 4+_logTable[3] == 4+1 = 5. On the other hand 2^14 gets your interval 7. So we never end up using interval 6. I guess I'm asking if that matters.

Reply:
2^14 - 1 is 0x7FFF (not 0x3FFF) so this will map to interval 6, whereas all the big blocks over 2^14 get mapped to interval 7 (last one)

Reply:
2^14 == 1 << 14 == `b'0100 0000 0000 0000` == 0x4000, so 2^14 - 1 is 0x3FFF.

Reply:
You are right about 2^14-1 == 0x3FFF.

Let me go step by step to avoid any mistakes.
This is the intent of the code I have:
[128, 256) --> Interval 0
[256, 512) --> Interval 1
[512, 1K) --> Interval 2
[1K, 2K) --> Interval 3
[2K, 4K) --> Interval 4
[4K, 8K) --> Interval 5
[8K, 16K) --> Interval 6
Everything larger or equal to 16K --> Interval 7

2^14 - 1 falls in the [8K, 16K) range so it should yield interval 6

2^14-1 = `b'11 1111 1111 1111` (that's 14 of ones)
After `blockSize >>= BITS_TO_SHIFT_FIRST_INTERVAL;` operation, block size removes 7 bits and we are left with just 7 of ones: `b'111 1111`.
Then we get into `(upper = blockSize >> 4) ? 4 + _logTable[upper] : _logTable[blockSize];`
`blockSize` is shifted 4 more positions so we are left with `111`. The lookup table is consulted for position 7 which yields  value 2. This added to 4 to yield interval 6.

Reply:
Ah, I see what I was doing wrong; I was right shifting by 8 (ie `NUM_INTERVALS`) instead of `BITS_TO_SHIFT_FIRST_INTERVAL` which is a right shift by 7. It's worth noting that we'll never actually read from `_logTable[0]`.